### DISCUSSION TITLE
Welcome to ikvm Discussions!

### DISCUSSION THREAD

**Discussion by wasabii** [ikvmnet/ikvm#129] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by yuzd** [comment]
I have an open source project about app deploying.Currently only supports dotnet applications.
https://github.com/yuzd/AntDeploy

I want to support java applications in my project,
 so I need a stripped down version of maven that doesn't depend on the java environment.
can ikvm-maven can help me do this?

**Reply by NightOwl888** [reply]
At present, parts of the Maven Resolver API are bundled into the [IKVM.Maven.Sdk](https://www.nuget.org/packages/IKVM.Maven.Sdk) NuGet package but aren't meant to be consumed directly by users. However, no objections if you want to contribute to help make Maven into its own NuGet package (and its own versioning cycle that is consistent with Maven).

Although we generally discourage people from uploading packages to NuGet.org with IKVM-converted assemblies in them, Maven Resolver is one exception that I would argue for because of its usefulness and the need to decouple it from `MavenReference` so the SDK can depend on it. Having a .NET version of it readily available also prevents the chicken and egg problem with compiling it to .NET in the first place.

That said, we don't currently use the "stripped down" set of minimal APIs that we consume. In particular, we can get by without the [`maven-core` package](https://github.com/ikvm-revived/ikvm-maven/blob/0d23ad7cf08b950fb235df5802d8b5410069e3b3/src/IKVM.Maven.Sdk.Tasks/ikvm-maven-sdk-tasks.pom#L45-L49) which is about 2MB extra (we could have 7MB instead of 9MB).

To get a look at how to set this up in a project, you need to start out with a Java JDK and Maven installed so you can do the initial Maven build. The command to download the `.jar` dependencies can be found in [IKVM.Maven.Sdk.Tasks.csproj](https://github.com/ikvm-revived/ikvm-maven/blob/0d23ad7cf08b950fb235df5802d8b5410069e3b3/src/IKVM.Maven.Sdk.Tasks/IKVM.Maven.Sdk.Tasks.csproj#L21).

Once they are downloaded they can be compiled with `ikvmc` or [`IkvmReference`](https://github.com/ikvm-revived/ikvm-maven/blob/0d23ad7cf08b950fb235df5802d8b5410069e3b3/src/IKVM.Maven.Sdk.Tasks/IKVM.Maven.Sdk.Tasks.csproj#L26-L30) into one or more assemblies. We currently compile it all into 1 assembly.

To use the compiled .NET Maven Resolver API in a project to download `.jar` files, have a look at [MavenReferenceItemResolve](https://github.com/ikvm-revived/ikvm-maven/blob/0d23ad7cf08b950fb235df5802d8b5410069e3b3/src/IKVM.Maven.Sdk.Tasks/MavenReferenceItemResolve.cs) and [MavenEnvironment](https://github.com/ikvm-revived/ikvm-maven/blob/0d23ad7cf08b950fb235df5802d8b5410069e3b3/src/IKVM.Maven.Sdk.Tasks/MavenEnvironment.cs).

Note that there are some more [simplified resolver API demos](https://github.com/apache/maven-resolver/tree/master/maven-resolver-demos/maven-resolver-demo-snippets/src/main/java/org/apache/maven/resolver/examples) and you can see the [minimum dependencies to use the Resolver API here](https://github.com/apache/maven-resolver/blob/master/maven-resolver-demos/maven-resolver-demo-snippets/pom.xml#L43-L142). (note that I had to add `commons-logging-1.2.jar` to this set in order to prevent errors in .NET).

**Comment by wasabii** [comment]
I strongly disagree. There is no need to decouple it. It's intended to be a private implementation.

As such you can add a Maven Reference to Maven

**Comment by kevincrans** [comment]
I've got a question, is or will J2ME applications ever be supported?
For debugging an app, I want a better environment than the Android J2ME emulator.

**Comment by wasabii** [comment]
This is not a focus, or even interest, of anybody I'm aware of, so probably never.

**Reply by kevincrans** [reply]
Thank you! I saw there are j2me emulators for java, so I'll test it.
